core: implement `DeterministicRandomSource`

joboet opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/131607)*
<!-- homu-ignore:end -->

ACP: rust-lang/libs-team#394
Tracking issue: rust-lang/rust#131606

The version implemented here uses ChaCha8 as RNG. Whether this is the right choice is still open for debate, so I've included the RNG name in the feature gate to make sure people need to recheck their code if we change the RNG.

Also, I've made one minor change to the API proposed in the ACP: in accordance with [C-GETTER](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter), `get_seed` is now named `seed`.

r? @joshtriplett

joboet committed:
core: implement `DeterministicRandomSource`

ACP: rust-lang/libs-team#394
Tracking issue: #131606

The version implemented here uses ChaCha8 as RNG. Whether this is the right choice is still open for debate, so I've included the RNG name in the feature gate to make sure people need to recheck their code if we change the RNG.

Also, I've made one minor change to the API proposed in the ACP: in accordance with [C-GETTER](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter), `get_seed` is now named `seed`.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11306690494/job/31447571627) [(plain)](https://github.com/rust-lang/rust/commit/94633268a834cddfff976f1c11a565865d571cd1/checks/31447571627/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
For more information about this error, try `rustc --explain E0658`.
error: test failed, to rerun pass `-p core --test coretests`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/cargo-miri runner /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/miri/x86_64-unknown-linux-gnu/debug/deps/coretests-07ae7ae7f436856e notest --quiet -Z unstable-options --format json` (exit status: 1)
note: test exited abnormally; to see the full output pass --nocapture to the harness.
  local time: Sat Oct 12 16:03:05 UTC 2024
  network time: Sat, 12 Oct 2024 16:03:05 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Noratrieb commented on library/core/src/random/deterministic.rs:
is this actually a guarantee we want in the standard library? I feel like this is a mistake. it locks the precise algorithm forever, even if there are future better ones. this is something a crate should do, which can do major versions to change it anyways, but for the standard library this sounds like a really bad idea.
C++ did this and is now stuck with a terrible rng.

juntyr replied:
That's a good point. As an ecosystem, we want to provide both a PRNG that is perfectly stable and never changes, and give those who accept breaks in RNG stability the option to upgrade. In std, this could be done by having >1 PRNGs, which would increase the maintenance burden over time. Having an external crate is a lot cleaner (some people can stick to v1, some to v2, and some always upgrade to the latest version).

It would probably still make sense to develop this crate in-house. Then, rust-lang/prand could release breaking versions when a breakthrough in PRNGs is made, Rust std could hide the latest prand implementation behind a newtype (similar to hashbrown for hash maps), and anyone who needs endless stability could depend on a pinned version of rust-lang/prand directly

joshtriplett replied:
> is this actually a guarantee we want in the standard library? I feel like this is a mistake. it locks the precise algorithm forever, even if there are future better ones. this is something a crate should do, which can do major versions to change it anyways, but for the standard library this sounds like a really bad idea. C++ did this and is now stuck with a terrible rng.
> 
> Being the same on all platforms sounds good though.

This was exactly the intent of the ACP, and the reason why we don't declare the RNG to be secure: we're prioritizing reproducibility.

Noratrieb commented on library/core/src/random/deterministic.rs:
while it could always be optimized in the future, I think it should at least be optimized (with SIMD) before stabilization. it would be quite unfortunate to land an rng in std that's slower than the rand crate, I think.

hanna-kruppe replied:
On that note, it may be a good idea to borrow two tweaks from [chacha8rand](https://github.com/C2SP/C2SP/blob/main/chacha8rand.md): defining the output to interleave multiple ChaCha20/8 blocks in the way 128-bit SIMD naturally does, and only adding the key to the final state matrix without also adding constants/counters/nonce to the other parts of the matrix. Both tweaks change the output, so they can't be done after stabilization, but they don't affect the *quality* and make SIMD implementations a little faster and simpler.

hanna-kruppe replied:
FWIW, I finally published [my implementation of chacha8rand](https://github.com/hanna-kruppe/chacha8rand) yesterday. I'm not suggesting that core should use this exact algorithm, let alone my implementation of it, but it may be useful as a reference for fast ChaCha8 SIMD implementations (including the aforementioned tweaks). Besides the portable scalar implementation I wrote SSE2, AVX2, AArch64 NEON, and wasm simd128 backends - much more platform coverage than `rand_chacha` currently has. The code is simpler and more self-contained than the `chacha20` crate because it doesn't need to be parametrized over number of rounds and algorithm variants or integrate with crypto traits/infrastructure.

juntyr commented:
If std provides a seeded PRNG, I think it should be one that can support reseeding, fast skipping, and forking in its API.

joshtriplett commented:
Chacha supports random access /seeking, and you can always replace the seed.

joshtriplett commented:
> Also, I've made one minor change to the API proposed in the ACP: in accordance with [C-GETTER](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter), `get_seed` is now named `seed`.

Seems reasonable.

hanna-kruppe commented on library/core/src/random/deterministic.rs:
This can throw away quite a bit of generated data, depending on the granularity of `fill_bytes` calls. When generating a `u32` via its `Random` impl, it generates one block of 64 bytes and throws away 60 of them. That's terrible for performance no matter how much SIMD you throw at it. It's better to keep a buffer of one or several blocks, incrementally consume the buffered bytes, and only refill once they're all used up. The logic for that is slightly more fiddly but it's absolutely worth it. That's another "performance optimization" that needs to be done before stabilization because it changes observable behavior.

joshtriplett replied:
That's something we probably wouldn't want to do if this was a secure generator, but since this is optimizing for performance and reproducibility, it seems reasonable.

Lokathor commented:
1) Since the generator is supposed to have a stable output sequence, should we name the generator after the generation method, allowing for other generation methods/sequences to be added in the future?
2) None of the type docs appear to state the generator's period before the sequence repeats. Can that be given somewhere?

programmerjake commented:
can accessors be added for `counter_nonce` and any byte buffer that might be added? this would allow seeking and [de]serializing the state.

joshtriplett commented:
> can accessors be added for `counter_nonce` and any byte buffer that might be added? this would allow seeking and [de]serializing the state.

I think that should happen in a second pass. And ideally we should encapsulate all the state in a structure rather than requiring users to request and pass several separate values.

joshtriplett commented:
@traviscross Any thoughts about the proposed switch to chacha8rand? It seems reasonable to me, but additional scrutiny would help.

hanna-kruppe commented:
Note that chacha8rand does not support seeking, which is sometimes desirable. You can store the current position in the byte stream compactly, but you can’t seek forward or backwards to arbitrary positions by just adjusting a block counter. However, the output-affecting optimizations (interleaving several chacha8 blocks, skipping addition of constants/counters/nonce to the final state) introduced by chacha8rand are interesting even if seeking is desired.

hanna-kruppe commented:
Right, that's what you'd do to get seeking. It just wouldn't be chacha8rand and should be called something different to avoid confusion.

hanna-kruppe commented:
[ChaCha4 is known to fail PractRand](https://www.johndcook.com/blog/2020/02/22/chacha-rng-with-fewer-rounds/) so it’s *worse* than the many PRNGs that pass it with smaller state size, smaller code size, similar or better performance on CPUs without SIMD, and less chance of being confused for something secure.

ChaCha6 may be acceptable for statistical purposes but the performance advantage over eight rounds is small. For long-output throughput, the maximum theoretical speed up is 8/6 = 1.33x, but this ignores the round-independent costs of per-block setup/finalization, internal buffer management, and actually doing something with the output. For shorter outputs such as single integers, the speedup will be even less because consuming a buffer in four or eight byte increments is a significant chunk of the work when filling the buffer takes less than a cycle per byte. For example, when I benchmark my chacha8rand implementation, the AVX2 backend is 2x faster than the SSE2 backend for reading several hundred bytes at a time but (from memory) only about 1.3x faster at reading individual u32s.

Reducing rounds also doesn’t address the drawbacks of generating output in multiples of 64 bytes (some implementation complexity and increased state size if you buffer internally; terrible performance for small reads if you don’t). If ChaCha8 is considered overkill then the logical conclusion to me is an entirely different algorithm that isn’t block based, not fewer rounds.

hanna-kruppe commented:
> I'd probably prefer to use a natively wider block permutation, and thereby benefit from diffusion across all words, rather than adapt ChaCha by striping across multiple instances in an ad-hoc way.

Is there such an algorithm today? It’s not clear to me that you can achieve similar or better diffusion across a larger state size with fewer operations *per byte/word*. Plus, a larger block size means more live values to keep in registers, which disproportionately penalizes processors without 128+ bit SIMD. ISAs with 16+ general purpose registers can compute one ChaCha block at a time without spilling to the stack (or only very occasionally with 16 GPRs), but that’s not true when a single block is 4x larger.

ChrisDenton commented:
Sorry if I've missed it but is there a particular reason we would prefer to use ChaCha? I mean, it makes sense in the context of this PR as a kind of placeholder for whatever the final algorithm is. But otherwise it seems like it's not really worth getting in the weeds over unless there isn't a more obviously fitting choice?

hanna-kruppe commented:
It's an unresolved question in the tracking issue, there was a little discussion about it there and much more on the ACP. I don't get the impression that there's any consensus about it. Speaking only for myself, I did not see any alternative suggestion that's *obviously better* choice than some suitable ChaCha variant. It's just a well understood option in the space, and has enough knobs that you can get easily get into the weeds over.

joboet committed:
core: cache the RNG output

joboet committed:
core: add RNG tests

joboet commented:
I've now implemented buffering and added some tests. As for other algorithms, I'd prefer it if that discussion could be moved to the [tracking issue](https://github.com/rust-lang/rust/issues/131606)?

hanna-kruppe commented on library/core/src/random/deterministic.rs:
```diff
@@ -117,6 +117,9 @@ pub struct DeterministicRandomSource {
     counter_nonce: u128,
+    block: [u8; 64],
+    // The amount of bytes in block that were already used.
+    used: u8,
```
While `u8` has sufficient range, it leads to a lot of casts in `fill_bytes` and `usize` is arguably more idiomatic for anything used for indexing/slicing. A potential reason to prefer `u8` would be a smaller memory footprint, but in this case it doesn't make a difference because of the alignment of `u128` (platform-specific but generally >= that of `usize`) and the nice round sizes of the other fields conspire to turn the "saved" bytes into padding.

hanna-kruppe replied:
Most 32- and 64-bit architectures don't have native 8-bit registers or operations on them. Even when they do (x86's vestigial `al`, `bl`, `cl`, etc.), they're never any *cheaper* in any way that would be applicable here (e.g., if you operate on a lot of values at once, smaller types are nice for SIMD, but here we only have one counter). The best you can hope for is that the sign/zero extensions and truncations can be mostly optimized out and don't cost anything extra on loads/stores, which is often but not always the case.

The only exception are some odd "8 bit ALU with 16 bit address space" architectures of which only AVR has any semblance of Rust support.

hanna-kruppe replied:
A very simple example for a bunch of targets: https://rust.godbolt.org/z/fYTqhqboq

hanna-kruppe replied:
I don't follow. I'm saying that sticking with `u8` here is *never* "a bit faster" (on any target worth worrying about), it's either a bit slower or performs exactly the same.

Lokathor commented:
As the cited article says: If it's not cryptographic to begin with, why chacha of any round count at all?

hanna-kruppe commented:
> The citation given does not support this claim. It says that ChaCha4 does pass Practrand (but that ChaCha3 does not):

You're right, sorry for the confusion. I might have misread the "up to 512 GB" bit in the footnote (as "... but fails with more output").

joboet committed:
core: use a `const` for the block size, use `usize` instead of `u8`

joshtriplett commented on library/core/src/random/deterministic.rs:
```suggestion
/// * Does *not* reseed on `fork`, VM fork, or in similar scenarios, meaning the
```

joshtriplett commented on library/core/tests/random.rs:
```diff
@@ -0,0 +1,82 @@
+    rng2.fill_bytes(&mut b2);
+    assert_eq!(b1, b2);
+
+    // Test the output of the RNG.
```
```suggestion
    // Test the output of the RNG.
    // If this test fails, this is a bug in the RNG, not the test.
    // Do not change the generated bytes below.
```

joboet committed:
core: improve comments

Co-authored-by: Josh Triplett <josh@joshtriplett.org>

bors commented:
:umbrella: The latest upstream changes (presumably #135937) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joboet commented:
I no longer think that this is a good idea; so I don't want to work on it further. If libs-api wants this regardless, feel free to use whatever is useful.
